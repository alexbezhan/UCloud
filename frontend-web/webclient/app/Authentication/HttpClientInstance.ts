import HttpClient from "./lib";
import {WebSocketFactory} from "./ws";

export const Client = new HttpClient();
export const WSFactory = new WebSocketFactory(Client);
