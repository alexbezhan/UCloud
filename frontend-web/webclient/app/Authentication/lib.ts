import * as jwt from "jsonwebtoken";
import {
    inDevEnvironment,
    inRange,
    inSuccessRange,
    is5xxStatusCode
} from "UtilityFunctions";
import {ReduxObject} from "DefaultObjects";
import {Store} from "redux";
import {SnackType} from "Snackbar/Snackbars";
import {snackbarStore} from "Snackbar/SnackbarStore";

export interface Override {
    path: string,
    method: { value: string }
    destination: {
        scheme?: string
        host?: string
        port: number
    }
}

interface CallParameters {
    method: string;
    path: string;
    body?: object;
    context?: string;
    maxRetries?: number;
    disallowProjects?: boolean;
}

/**
 * Represents an instance of the SDUCloud object used for contacting the backend, implicitly using JWTs.
 */
export default class SDUCloud {
    private readonly context: string;
    private readonly serviceName: string;
    private readonly authContext: string;
    private readonly redirectOnInvalidTokens: boolean;

    private apiContext: string;
    private accessToken: string;
    private csrfToken: string;
    private decodedToken: any;
    private forceRefresh: boolean = false;
    private overridesPromise: Promise<void> | null = null;

    private projectId: string | undefined = undefined;
    private projectAccessToken: string | undefined = undefined;
    private projectDecodedToken: any | undefined = undefined;

    overrides: Override[] = [];

    constructor() {
        let context = location.protocol + '//' +
            location.hostname +
            (location.port ? ':' + location.port : '');

        let serviceName: string;
        switch (location.hostname) {
            case "localhost":
                serviceName = "local-dev-csrf";
                break;

            default:
                serviceName = "web-csrf";
                break;
        }

        this.context = context;
        this.serviceName = serviceName;

        this.apiContext = "/api";
        this.authContext = "/auth";

        this.decodedToken = null;
        this.redirectOnInvalidTokens = false;

        let accessToken = SDUCloud.storedAccessToken;
        let csrfToken = SDUCloud.storedCsrfToken;
        if (accessToken && csrfToken) {
            this.setTokens(accessToken, csrfToken);
        }

        if (process.env.NODE_ENV === "development") {
            this.overridesPromise = (async () => {
                const jsonResponse: Override[] = await (await fetch("http://localhost:9900/")).json();
                console.log("Got back the following response:", jsonResponse);
                this.overrides = jsonResponse;
            })();
        }
    }

    initializeStore(store: Store<ReduxObject>) {
        store.subscribe(() => {
            const project = store.getState().project.project;
            if (project !== this.projectId) {
                console.log("Something");
                this.projectId = project;
                this.projectDecodedToken = undefined;
                this.projectAccessToken = undefined;
            }
        });
    }

    /**
     * Makes an AJAX call to the API. This will automatically add relevant authorization headers.
     * If the user's JWT has expired this will automatically attempt to refresh it.
     *
     * The path argument should be without the context, it should also not include the /api/ part of the context.
     * For example, to call `GET https://cloud.sdu.dk/api/files?path=/home/foobar` you should use
     * `cloud.call("GET", "/files?path=/home/foobar/")`
     *
     * The body argument is assumed to be JSON.
     *
     * @param {string} method - the HTTP method
     * @param {string} path - the path, should not contain context or /api/
     * @param {object} body - the request body, assumed to be a JS object to be encoded as JSON.
     * @param {string} context - the base of the request (e.g. "/api")
     * @param {number} maxRetries - the amount of times the call should be retried on failure (Default: 5).
     * @param {disallowProjects} disallowProjects - true if this call should not use the project token (Default: false).
     * @return {Promise} promise
     */
    async call(
        {
            method,
            path,
            body,
            context = this.apiContext,
            maxRetries = 5,
            disallowProjects = false
        }: CallParameters
    ): Promise<any> {
        if (this.overridesPromise != null) {
            try {
                await this.overridesPromise;
            } catch (ignored) {
            }
            this.overridesPromise = null;
        }

        if (path.indexOf("/") !== 0) path = "/" + path;
        return this.receiveAccessTokenOrRefreshIt(disallowProjects).then(token => {
            return new Promise((resolve, reject) => {
                let req = new XMLHttpRequest();
                req.open(method, this.computeURL(context, path));
                req.setRequestHeader("Authorization", `Bearer ${token}`);
                req.setRequestHeader("Content-Type", "application/json; charset=utf-8");
                req.responseType = "text"; // Explicitly set, otherwise issues with empty response

                const rejectOrRetry = (parsedResponse?) => {
                    if (req.status === 401) {
                        this.forceRefresh = true;
                    }

                    if (maxRetries >= 1 && is5xxStatusCode(req.status)) {
                        this.call({
                            method: method,
                            path: path,
                            body: body,
                            context: context,
                            maxRetries: maxRetries - 1
                        })
                            .catch(e => reject(e)).then(e => resolve(e));
                    } else {
                        reject({request: req, response: parsedResponse});
                    }
                };

                req.onload = async () => {
                    try {
                        let responseContentType = req.getResponseHeader("content-type");
                        let parsedResponse = req.response.length === 0 ? "{}" : req.response;

                        // JSON Parsing
                        if (responseContentType !== null) {
                            if (responseContentType.indexOf("application/json") !== -1 ||
                                responseContentType.indexOf("application/javascript") !== -1) {
                                parsedResponse = JSON.parse(parsedResponse);
                            }
                        }

                        if (inSuccessRange(req.status)) {
                            resolve({
                                response: parsedResponse,
                                request: req,
                            });
                        } else {
                            rejectOrRetry(parsedResponse);
                        }
                    } catch (e) {
                        rejectOrRetry();
                    }
                };

                if (body) {
                    req.send(JSON.stringify(body));
                } else {
                    req.send();
                }
            });
        });
    }

    public computeURL(context: string, path: string): string {
        let absolutePath = context + path;
        for (let i = 0; i < this.overrides.length; i++) {
            let override = this.overrides[i];
            if (absolutePath.indexOf(override.path) === 0) {
                let scheme = override.destination.scheme ?
                    override.destination.scheme : "http";
                let host = override.destination.host ?
                    override.destination.host : "localhost";
                let port = override.destination.port;

                return scheme + "://" + host + ":" + port + absolutePath;
            }
        }

        return this.context + absolutePath;
    }

    /**
     * Calls with the GET HTTP method. See call(method, path, body)
     */
    public async get<T = any>(path: string, context = this.apiContext,
                              disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "GET", path, body: undefined, context, disallowProjects});
    }

    /**
     * Calls with the POST HTTP method. See call(method, path, body)
     */
    public async post<T = any>(path: string, body?: object, context = this.apiContext,
                               disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "POST", path, body, context, disallowProjects});
    }

    /**
     * Calls with the PUT HTTP method. See call(method, path, body)
     */
    public async put<T = any>(path: string, body: object, context = this.apiContext,
                              disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "PUT", path, body, context, disallowProjects});
    }

    /**
     * Calls with the DELETE HTTP method. See call(method, path, body)
     */
    public async delete<T = void>(path: string, body: object, context = this.apiContext,
                                  disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "DELETE", path, body, context, disallowProjects});
    }

    /**
     * Calls with the PATCH HTTP method. See call(method, path, body)
     */
    public async patch<T = any>(path: string, body: object, context = this.apiContext,
                                disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "PATCH", path, body, context, disallowProjects});
    }

    /**
     * Calls with the OPTIONS HTTP method. See call(method, path, body)
     */
    public async options<T = any>(path: string, body: object, context = this.apiContext,
                                  disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "OPTIONS", path, body, context, disallowProjects});
    }

    /**
     * Calls with the HEAD HTTP method. See call(method, path, body)
     */
    public async head<T = any>(path: string, context = this.apiContext,
                               disallowProjects: boolean = false): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call({method: "HEAD", path, body: undefined, context, disallowProjects});
    }

    /**
     * Opens up a new page which contains the login page at the auth service. This login page will automatically
     * redirect back to the correct service (using serviceName).
     */
    public openBrowserLoginPage() {
        if (window.location.href !== this.context + "/app/login")
            window.location.href = this.context + "/app/login";
    }

    public openLandingPage() {
        if (window.location.href !== this.context + "/app/")
            window.location.href = this.context + "/app/";
    }

    /**
     * @returns the username of the authenticated user or null
     */
    public get username(): string | undefined {
        let info = this.userInfo;
        if (info) return info.sub;
        else return undefined;
    }

    public get activeUsername(): string | undefined {
        if (this.useProjectToken(false) && !!this.projectDecodedToken) {
            return this.projectDecodedToken.payload.sub;
        } else {
            return this.username;
        }
    }

    /**
     * @returns {string} the homefolder path for the currently logged in user (with trailing slash).
     */
    public get homeFolder(): string {
        let username = this.username;
        if (this.projectId !== undefined) {
            username = this.projectId;
        }
        return `/home/${username}/`
    }

    public get trashFolder(): string {
        return `${this.homeFolder}Trash/`
    }

    public get isLoggedIn(): boolean {
        return this.userInfo != null;
    }

    /**
     * @returns {string} the userrole. Null if none available in the JWT
     */
    get userRole(): string {
        const info = this.userInfo;
        if (info) return info.role;
        return "";
    }

    /**
     * @returns {boolean} whether or not the user is listed as an admin.
     */
    get userIsAdmin(): boolean {
        return this.userRole === "ADMIN";
    }

    /**
     * Returns the userInfo (the payload of the JWT). Be aware that the JWT is not verified, this means that a user
     * will be able to put whatever they want in this. This is normally not a problem since all backend services _will_
     * verify the token.
     */
    private get userInfo(): undefined | JWT {
        let token = this.decodedToken;
        if (!token) return undefined;
        else return this.decodedToken.payload;
    }

    get principalType(): undefined | string {
        const userInfo = this.userInfo;
        if (!userInfo) return undefined;
        else return userInfo.principalType;
    }

    private useProjectToken(disallowProjects: boolean): boolean {
        return this.projectId !== undefined && !disallowProjects;
    }

    private retrieveToken(disallowProjects: boolean): string {
        if (this.useProjectToken(disallowProjects)) {
            return this.projectAccessToken!;
        } else {
            return SDUCloud.storedAccessToken;
        }
    }

    /**
     * Attempts to receive a (non-expired) JWT access token from storage. In case the token has expired at attempt will
     * be made to refresh it. If it is not possible to refresh the token a MissingAuthError will be thrown. This would
     * indicate the user no longer has valid credentials. At this point it would make sense to present the user with
     * the login page.
     *
     * @return {Promise} a promise of an access token
     */
    async receiveAccessTokenOrRefreshIt(disallowProjects: boolean = false): Promise<any> {
        if (this.overridesPromise != null) {
            try {
                await this.overridesPromise;
            } catch (ignored) {
            }
            this.overridesPromise = null;
        }

        let tokenPromise: Promise<any> | null = null;
        if (this.isTokenExpired(disallowProjects) || this.forceRefresh) {
            tokenPromise = this.refresh(disallowProjects);
            this.forceRefresh = false;
        } else {
            tokenPromise = new Promise((resolve, reject) => resolve(this.retrieveToken(disallowProjects)));
        }
        return tokenPromise;
    }

    createOneTimeTokenWithPermission(permission, disallowProjects: boolean = false): Promise<any> {
        return this.receiveAccessTokenOrRefreshIt(disallowProjects)
            .then(token => {
                let oneTimeToken = `${this.context}${this.authContext}/request/?audience=${permission}`;
                return new Promise((resolve, reject) => {
                    let req = new XMLHttpRequest();
                    req.open("POST", oneTimeToken);
                    req.setRequestHeader("Authorization", `Bearer ${token}`);
                    req.setRequestHeader("Content-Type", "application/json");
                    req.onload = () => {
                        try {
                            if (inRange({status: req.status, min: 200, max: 299})) {
                                const response = req.response.length === 0 ? "{}" : req.response;
                                resolve({response: JSON.parse(response), request: req});
                            } else {
                                reject(req.response);
                            }
                        } catch (e) {
                            reject(e.response)
                        }
                    };
                    req.send();
                });
            }).then((data: any) => new Promise((resolve, reject) => resolve(data.response.accessToken)));
    }

    private async refresh(disallowProjects: boolean): Promise<string> {
        const project = this.projectId;
        if (project !== undefined && !disallowProjects) {
            const result = await this.post("/projects/auth", {project}, undefined, true);
            if (inSuccessRange(result.request.status)) {
                const accessToken = result.response.accessToken;
                this.projectAccessToken = accessToken;
                this.projectDecodedToken = this.decodeToken(accessToken);
                return accessToken;
            } else {
                if (result.request.status === 401 || result.request.status === 400) {
                    SDUCloud.clearTokens();
                    this.openBrowserLoginPage();
                }

                throw Error("Unable to refresh token")
            }
        } else {
            let csrfToken = SDUCloud.storedCsrfToken;
            if (!csrfToken) {
                return new Promise((resolve, reject) => {
                    reject(this.missingAuth());
                });
            }

            let refreshPath = this.computeURL(this.authContext, "/refresh/web");
            return new Promise((resolve, reject) => {
                let req = new XMLHttpRequest();
                req.open("POST", refreshPath);
                req.setRequestHeader("X-CSRFToken", csrfToken);
                req.onload = () => {
                    try {
                        if (inSuccessRange(req.status)) {
                            resolve(JSON.parse(req.response));
                        } else {
                            if (req.status === 401 || req.status === 400) {
                                SDUCloud.clearTokens();
                                this.openBrowserLoginPage();
                            }
                            reject({status: req.status, response: req.response});
                        }
                    } catch (e) {
                        reject({status: e.status, response: e.response});
                    }
                };
                req.send();
            }).then((data: any) => {
                return new Promise((resolve, reject) => {
                    this.setTokens(data.accessToken, data.csrfToken);
                    resolve(data.accessToken);
                });
            });
        }
    }

    /**
     * Updates tokens received by the auth service.
     *
     * @param accessToken the (JWT) access token.
     * @param csrfToken the csrf token
     */
    public setTokens(accessToken: string, csrfToken: string): void {
        if (!accessToken) throw this.missingAuth();

        this.accessToken = accessToken;
        SDUCloud.storedAccessToken = accessToken;

        this.csrfToken = csrfToken;
        SDUCloud.storedCsrfToken = csrfToken;


        this.decodedToken = this.decodeToken(accessToken);
    }

    private decodeToken(accessToken: string): any {
        const bail = (): never => {
            SDUCloud.clearTokens();
            this.openBrowserLoginPage();
            return void (0) as never;
        };
        try {
            const token = jwt.decode(accessToken, {complete: true});

            if (token === null) {
                return bail();
            }

            if (typeof token === "string") {
                return bail();
            } else if (typeof token === "object") {
                const payload = token.payload;
                const isValid = "sub" in payload &&
                    "uid" in payload &&
                    "aud" in payload &&
                    "role" in payload &&
                    "iss" in payload &&
                    "exp" in payload &&
                    "extendedByChain" in payload &&
                    "iat" in payload &&
                    "principalType" in payload;

                if (!isValid) {
                    console.log("Bailing. Bad JWT");
                    return bail();
                }

                return token;
            } else {
                return bail();
            }
        } catch (e) {
            return bail();
        }
    }

    public async logout() {
        try {
            const res = await fetch(`${this.context}${this.authContext}/logout/web`, {
                headers: {
                    "X-CSRFToken": SDUCloud.storedCsrfToken,
                    "Content-Type": "application/json",
                },
                method: "POST",
                credentials: "same-origin"
            });
            if (!is5xxStatusCode(res.status)) {
                window.localStorage.removeItem("accessToken");
                window.localStorage.removeItem("csrfToken");
                this.openBrowserLoginPage();
                return;
            }
            throw Error("The server was unreachable, please try again later.")
        } catch (err) {
            snackbarStore.addSnack({message: err.message, type: SnackType.Failure});
        }
    }

    private static clearTokens() {
        SDUCloud.storedAccessToken = "";
        SDUCloud.storedCsrfToken = "";
    }

    static get storedAccessToken(): string {
        return window.localStorage.getItem("accessToken") || "";
    }

    static set storedAccessToken(value: string) {
        window.localStorage.setItem("accessToken", value);
    }

    static get storedCsrfToken(): string {
        return window.localStorage.getItem("csrfToken") || "";
    }

    static set storedCsrfToken(value) {
        window.localStorage.setItem("csrfToken", value);
    }

    private isTokenExpired(disallowProject: boolean) {
        const token = this.useProjectToken(disallowProject) ? this.projectDecodedToken : this.decodedToken;
        if (!token || !token.payload) return true;
        const nowInSeconds = Math.floor(Date.now() / 1000);
        const inOneMinute = nowInSeconds + (60);
        return token.payload.exp < inOneMinute;
    }

    private missingAuth() {
        if (this.redirectOnInvalidTokens) {
            this.openBrowserLoginPage();
            return 0;
        } else {
            return new MissingAuthError();
        }
    }
}

interface JWT {
    sub: string
    uid: number
    lastName?: string
    aud: string
    role: string
    iss: string
    firstNames?: string
    exp: number
    extendedByChain: any[]
    iat: number
    principalType: string
    publicSessionReference?: string
}

export class MissingAuthError {
    private name: string;

    constructor() {
        this.name = "MissingAuthError";
    }
}
