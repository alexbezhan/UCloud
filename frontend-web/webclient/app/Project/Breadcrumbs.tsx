import * as React from "react";
import styled from "styled-components";
import Spinner, {HexSpinWrapper} from "LoadingIcon/LoadingIcon";
import {BreadCrumbsBase} from "ui-components/Breadcrumbs";
import {Link} from "ui-components";
import {useProjectManagementStatus} from "Project/index";

export interface ProjectBreadcrumbsProps {
    crumbs: {title: string, link?: string}[];
    allowPersonalProject?: true;
}

const ProjectBreadcrumbsWrapper = styled(BreadCrumbsBase)`
    width: 100%;
    max-width: unset;
    flex-grow: 1;
    
    ${HexSpinWrapper} {
        margin: 0;
        display: inline;
    }
`;

export const ProjectBreadcrumbs: React.FunctionComponent<ProjectBreadcrumbsProps> = props => {
    const {projectDetails, projectId} = useProjectManagementStatus({
        isRootComponent: false,
        allowPersonalProject: props.allowPersonalProject
    });
    let projectNameComponent = <Spinner />;
    if (!projectDetails.loading) {
        const title = projectDetails.data.title;
        projectNameComponent = <>{title.slice(0, 20).trim()}{title.length > 20 ? "..." : ""}</>;
    }

    return <ProjectBreadcrumbsWrapper mb="12px" embedded={false}>
        <span><Link to="/projects">My Projects</Link></span>
        {projectId ? (
            <span><Link to="/project/dashboard">{projectNameComponent}</Link></span>
        ) : props.allowPersonalProject ? <span><Link to="/project/dashboard">Personal Project</Link></span> : null}
        {props.crumbs.map((crumb, idx) => {
            if (crumb.link) {
                return <span key={idx}><Link to={crumb.link}>{crumb.title}</Link></span>;
            } else {

                return <span key={idx}>{crumb.title}</span>;
            }
        })}
    </ProjectBreadcrumbsWrapper>;
};
