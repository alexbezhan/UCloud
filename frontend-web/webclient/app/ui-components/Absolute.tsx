import styled from "styled-components";
import {
  bottom,
  BottomProps,
  left,
  LeftProps,
  right,
  RightProps,
  top,
  TopProps,
  zIndex,
  ZIndexProps
} from "styled-system";
import Box, {BoxProps} from "./Box";

interface AbsoluteProps extends BoxProps, TopProps, BottomProps, LeftProps, RightProps, ZIndexProps {
}

const Absolute = styled(Box) <AbsoluteProps>`
  position: absolute;
  ${top} ${bottom} ${left} ${right}
  ${zIndex}
`;

Absolute.displayName = "Absolute";

export default Absolute;
