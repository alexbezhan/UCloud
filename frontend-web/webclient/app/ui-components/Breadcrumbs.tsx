import * as React from "react";
import styled from "styled-components";
import {Box, Icon, Text} from "ui-components";
import {addTrailingSlash, removeTrailingSlash} from "UtilityFunctions";

// https://www.w3schools.com/howto/howto_css_breadcrumbs.asp
const BreadCrumbsBase = styled.ul`
    padding: 0;
    padding-right: 10px;
    margin: 0;
    list-style: none;
    max-width: 85%;
    height: 85px;
    overflow-y: auto;

    & li {
        display: inline;
        font-size: 25px;
    }

    & li + li:before {
        padding: 8px;
        color: var(--text, #f00);
        content: "/";
    }

    & li span {
        color: var(--text, #f00);
        text-decoration: none;
    }

    & li span:hover {
        cursor: pointer;
        color: var(--blue, #f00);
        text-decoration: none;
    }
`;

export interface BreadcrumbsList {
    currentPath: string;
    navigate: (path: string) => void;
    homeFolder: string;
    projectFolder: string;
}

export interface BreadCrumbMapping {
    actualPath: string;
    local: string;
}

export const BreadCrumbs = ({
    currentPath,
    navigate,
    homeFolder,
    projectFolder
}: BreadcrumbsList): JSX.Element | null => {
    if (!currentPath) return null;
    const pathsMapping = buildBreadCrumbs(currentPath, homeFolder, projectFolder);
    const activePathsMapping = pathsMapping[pathsMapping.length - 1];
    pathsMapping.pop();
    const breadcrumbs = pathsMapping.map((path, index) => (
        <li key={index}>
            <span title={path.local} onClick={() => navigate(path.actualPath)}>
                {`${path.local.slice(0, 20).trim()}${path.local.length > 20 ? "..." : ""}`}
            </span>
        </li>
    ));

    const addHomeFolderLink = !currentPath.startsWith(removeTrailingSlash(homeFolder));

    return (
        <>
            {addHomeFolderLink ? (
                <>
                    <Box ml="15px">
                        <Icon size="30px" cursor="pointer" name="home" onClick={toHome} />
                        <Text cursor="pointer" ml="-15px" fontSize="11px" onClick={toHome}>
                            Go to home
                        </Text>
                    </Box>
                    <Text ml="6px" mr="6px" fontSize="24px">|</Text>
                </>
            ) : null}
            <BreadCrumbsBase>
                {breadcrumbs}
                <li title={activePathsMapping.local}>
                    {activePathsMapping.local.slice(0, 20).trim()}{activePathsMapping.local.length > 20 ? "..." : ""}
                </li>
            </BreadCrumbsBase>
        </>
    );

    function toHome(): void {
        navigate(homeFolder);
    }
};

function buildBreadCrumbs(path: string, homeFolder: string, projectFolder: string): BreadCrumbMapping[] {
    const paths = path.split("/").filter(p => p !== "");
    if (paths.length === 0) return [{actualPath: "/", local: "/"}];

    const pathsMapping: BreadCrumbMapping[] = [];
    for (let i = 0; i < paths.length; i++) {
        let actualPath = "/";
        for (let j = 0; j <= i; j++) actualPath += paths[j] + "/";
        pathsMapping.push({actualPath, local: paths[i]});
    }

    // Handle starts with home
    if (addTrailingSlash(path).startsWith(homeFolder)) // remove first two indices
        return [{actualPath: homeFolder, local: "Home"}].concat(pathsMapping.slice(2));
    else if (path.startsWith("/home") && pathsMapping.length >= 2)
        return [{
            actualPath: pathsMapping[1].actualPath,
            local: `Home of ${pathsMapping[1].local}`
        }].concat(pathsMapping.slice(2));

    // Handle starts with project
    if (addTrailingSlash(path).startsWith("/projects/")) {
        return [{actualPath: projectFolder, local: "Projects"}].concat(pathsMapping.slice(2));
    }

    return pathsMapping;
}
