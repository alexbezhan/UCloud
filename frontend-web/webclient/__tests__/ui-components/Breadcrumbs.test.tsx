import {configure, mount} from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as React from "react";
import {create} from "react-test-renderer";
import {ThemeProvider} from "styled-components";
import {BreadCrumbs} from "../../app/ui-components/Breadcrumbs";
import theme from "../../app/ui-components/theme";
import {Client} from "../../app/Authentication/HttpClientInstance";
import {Provider} from "react-redux";
import {store} from "../../app/Utilities/ReduxUtilities";
configure({adapter: new Adapter()});


describe("Breadcrumbs", () => {
    it("Build breadcrumbs, embedded", () => {
        expect(create(
            <ThemeProvider theme={theme}>
                <Provider store={store}>
                    <BreadCrumbs
                        embedded
                        currentPath="/home/mail@mailhost.dk/folder1"
                        navigate={() => undefined}
                        client={Client}
                    />
                </Provider>
            </ThemeProvider>)).toMatchSnapshot();
    });

    it("Build breadcrumbs with empty path, embedded", () => {
        expect(create(
            <ThemeProvider theme={theme}>
                <Provider store={store}>
                    <BreadCrumbs
                        embedded
                        currentPath=""
                        navigate={() => undefined}
                        client={Client}
                    />
                </Provider>
            </ThemeProvider>
        )).toMatchSnapshot();
    });

    it("Build breadcrumbs, unembedded", () => {
        expect(create(
            <ThemeProvider theme={theme}>
                <Provider store={store}>
                    <BreadCrumbs
                        embedded={false}
                        currentPath="/home/mail@mailhost.dk/folder1"
                        navigate={() => undefined}
                        client={Client}
                    />
                </Provider>
            </ThemeProvider>)).toMatchSnapshot();
    });

    it("Build breadcrumbs with empty path, unembedded", () => {
        expect(create(
            <ThemeProvider theme={theme}>
                <Provider store={store}>
                    <BreadCrumbs
                        embedded={false}
                        currentPath=""
                        navigate={() => undefined}
                        client={Client}
                    />
                </Provider>
            </ThemeProvider>
        )).toMatchSnapshot();
    });

    it("Using navigate", () => {
        const navigate = jest.fn();
        const breadcrumbs = mount(
            <ThemeProvider theme={theme}>
                <Provider store={store}>
                    <BreadCrumbs
                        embedded
                        currentPath="/home/mail@mailhost.dk/folder1"
                        navigate={navigate}
                        client={Client}
                    />
                </Provider>
            </ThemeProvider>
        );
        breadcrumbs.find("span").first().simulate("click");
        expect(navigate).toHaveBeenCalled();
    });
});
