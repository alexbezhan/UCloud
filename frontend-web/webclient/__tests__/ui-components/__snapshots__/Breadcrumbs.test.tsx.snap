// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs Build breadcrumbs with empty path, embedded 1`] = `null`;

exports[`Breadcrumbs Build breadcrumbs with empty path, unembedded 1`] = `null`;

exports[`Breadcrumbs Build breadcrumbs, embedded 1`] = `
Array [
  .c0 {
  margin-left: 15px;
}

.c1 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: pointer;
}

.c2 {
  cursor: pointer;
  font-size: 11px;
  margin-left: -15px;
}

<div
    className="c0"
  >
    <svg
      className="c1"
      cursor="pointer"
      fill="currentcolor"
      height="30px"
      onClick={[Function]}
      test-tag="to_home"
      viewBox="0 0 148.167 127.257"
      width="30px"
    >
      <path
        d="M0 63.5l10.583 10.583 63.5-52.916 63.5 52.916L148.167 63.5 74.083 0z"
      />
      <path
        d="M74.083 28.149l-50.27 41.618v57.49h34.316V92.312c0-4.162 2.06-11.64 15.954-11.628 13.894.012 15.958 7.406 15.958 11.71v34.863h34.313v-57.49zM124.354 38.92l-19.553-17.224V5.252l19.553.04z"
      />
    </svg>
    <div
      className="c2"
      cursor="pointer"
      fontSize="11px"
      onClick={[Function]}
    >
      Go to home
    </div>
  </div>,
  .c0 {
  cursor: inherit;
  font-size: 24px;
  margin-left: 6px;
  margin-right: 6px;
}

<div
    className="c0"
    cursor="inherit"
    fontSize="24px"
  >
    |
  </div>,
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: calc(100% - 50px);
}

.c0 > span {
  width: 1;
  font-size: 25px;
  display: inline-block;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c0 > span + span:before {
  padding: 0 8px;
  vertical-align: top;
  color: var(--text,#f00);
  content: "/";
}

.c0 > span {
  color: var(--text,#f00);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 > span:hover {
  cursor: pointer;
  color: var(--blue,#f00);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 > span:last-child:hover {
  color: var(--text,#f00);
  cursor: default;
}

<div
    className="sc-fzpans c0"
  >
    <span
      onClick={[Function]}
      test-tag="Home of mail@mailhost.dk"
      title="Home of mail@mailhost.dk"
    >
      Home of mail@mailhost.dk
    </span>
    <span
      title="folder1"
    >
      folder1
    </span>
  </div>,
]
`;

exports[`Breadcrumbs Build breadcrumbs, unembedded 1`] = `
Array [
  .c0 {
  margin-left: 15px;
}

.c1 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: pointer;
}

.c2 {
  cursor: pointer;
  font-size: 11px;
  margin-left: -15px;
}

<div
    className="c0"
  >
    <svg
      className="c1"
      cursor="pointer"
      fill="currentcolor"
      height="30px"
      onClick={[Function]}
      test-tag="to_home"
      viewBox="0 0 148.167 127.257"
      width="30px"
    >
      <path
        d="M0 63.5l10.583 10.583 63.5-52.916 63.5 52.916L148.167 63.5 74.083 0z"
      />
      <path
        d="M74.083 28.149l-50.27 41.618v57.49h34.316V92.312c0-4.162 2.06-11.64 15.954-11.628 13.894.012 15.958 7.406 15.958 11.71v34.863h34.313v-57.49zM124.354 38.92l-19.553-17.224V5.252l19.553.04z"
      />
    </svg>
    <div
      className="c2"
      cursor="pointer"
      fontSize="11px"
      onClick={[Function]}
    >
      Go to home
    </div>
  </div>,
  .c0 {
  cursor: inherit;
  font-size: 24px;
  margin-left: 6px;
  margin-right: 6px;
}

<div
    className="c0"
    cursor="inherit"
    fontSize="24px"
  >
    |
  </div>,
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: calc(100% - 180px);
}

.c0 > span {
  width: 1;
  font-size: 25px;
  display: inline-block;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c0 > span + span:before {
  padding: 0 8px;
  vertical-align: top;
  color: var(--text,#f00);
  content: "/";
}

.c0 > span {
  color: var(--text,#f00);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 > span:hover {
  cursor: pointer;
  color: var(--blue,#f00);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 > span:last-child:hover {
  color: var(--text,#f00);
  cursor: default;
}

<div
    className="sc-fzpans c0"
  >
    <span
      onClick={[Function]}
      test-tag="Home of mail@mailhost.dk"
      title="Home of mail@mailhost.dk"
    >
      Home of mail@mailhost.dk
    </span>
    <span
      title="folder1"
    >
      folder1
    </span>
  </div>,
]
`;
